lib/bulk/common.js

```diff
@@ -427,26 +526,589 @@ function handleMongoWriteConcernError(batch, bulkResult, ordered, err, callback)
+    this.s.currentOp = null;
+    return this.s.options.addToOperationsList(this, UPDATE, document);
+  }
+
+  /**
+   * Add a replace one operation to the bulk operation
+   *
+   * @method
+   * @param {object} updateDocument the new document to replace the existing one with
+   * @throws {MongoError}
+   * @return {OrderedBulkOperation|UnorderedBulkOperation}
+   */
+  replaceOne(updateDocument) {
+    this.updateOne(updateDocument);
+  }
+
+  /**
+   * Upsert modifier for update bulk operation
+   *
+   * @method
+   * @throws {MongoError}
+   * @return {FindOperators}
+   */
+  upsert() {
+    this.s.currentOp.upsert = true;
+    return this;
+  }
+
+  /**
+   * Add a delete one operation to the bulk operation
+   *
+   * @method
+   * @throws {MongoError}
+   * @return {OrderedBulkOperation|UnordedBulkOperation}
+   */
+  deleteOne() {
+    // Establish the update command
+    const document = {
+      q: this.s.currentOp.selector,
+      limit: 1
+    };
+
+    // Clear out current Op
+    this.s.currentOp = null;
+    return this.s.options.addToOperationsList(this, REMOVE, document);
+  }
+
+  /**
+   * Add a delete operation to the bulk operation
+   *
+   * @method
+   * @throws {MongoError}
+   * @return {OrderedBulkOperation|UnordedBulkOperation}
+   */
+  delete() {
+    // Establish the update command
+    const document = {
+      q: this.s.currentOp.selector,
+      limit: 0
+    };
+
+    // Clear out current Op
+    this.s.currentOp = null;
+    return this.s.options.addToOperationsList(this, REMOVE, document);
+  }
+
+  /**
+   * backwards compatability for deleteOne
+   */
+  removeOne() {
+    return this.deleteOne();
+  }
+
+  /**
+   * backwards compatability for delete
+   */
+  remove() {
+    return this.delete();
+  }
+}
+
+/**
+ * Parent class to OrderedBulkOperation and UnorderedBulkOperation
+ * @class
+ */
+class BulkOperationBase {
+  /**
+   * Create a new OrderedBulkOperation or UnorderedBulkOperation instance (INTERNAL TYPE, do not instantiate directly)
+   * @class
+   * @property {number} length Get the number of operations in the bulk.
+   * @return {OrderedBulkOperation|UnordedBulkOperation}
+   */
+  constructor(topology, collection, options, isOrdered) {
+    // determine whether bulkOperation is ordered or unordered
+    this.isOrdered = isOrdered;
+
+    options = options == null ? {} : options;
+    // TODO Bring from driver information in isMaster
+    // Get the namespace for the write operations
+    const namespace = collection.collectionName;
+    // Used to mark operation as executed
+    const executed = false;
+
+    // Current item
+    const currentOp = null;
+
+    // Handle to the bson serializer, used to calculate running sizes
+    const bson = topology.bson;
+
+    // Set max byte size
+    const maxBatchSizeBytes =
+      topology.isMasterDoc && topology.isMasterDoc.maxBsonObjectSize
+        ? topology.isMasterDoc.maxBsonObjectSize
+        : 1024 * 1025 * 16;
```

Comment:
Where does `1024 * 1025 * 16` come from? It looks like `1025` might be a typo.

Reply:
I'm not sure where this number is determined. This line existed in unordered and ordered before the refactor. I can take out the 1025 if that is the decision.

Reply:
the spec has the maxBSONObjectSize as 16MB which would imply that the 1025 should be a 1024, but is that the value that we're looking for because there was nothing for matchBatchSizeBytes in the spec.
https://github.com/mongodb/specifications/blob/master/source/driver-bulk-update.rst#request-size-limits

Reply:
So I would expect it to be `1024 * 1024 * 16`, since the default max BSON size is 16 MB. `1024 * 1025 * 16` makes it 16MB + 16 KB, which doesn't really make sense